### PR TITLE
Prevent XSS attack using /auth/enable_cookies

### LINKED ISSUE #1453: /auth/enable_cookies XSS attack?
## Overview

/auth/enable_cookies appears to render arbitrary javascript - XSS

## Consuming repo

We have the shopify koa app mounted at: (real URL available in private email to seamus@abshere.net)

https://example.com/shopify

If you click this, you will get a "hello world" prompt.

https://example.com/shopify/auth/enable_cookies?shop=%27%22%3C/title%3E%3C/style%3E%3C/textarea%3E%3C/noscript%3E%3C/script%3E--%3E%3Cdtfy%3E%3Cscript%3Ealert(%27hello%20world%27)%3C/script%3E

## Labels

- [ ] Add the `Type: Bug` label to this issue.

### Area

- [ ] Add any relevant `Area: <area>` labels to this issue

### Scope

- Is this issue related to a specific package?

  - [ ] Tag it with the `Package: <package_name>` label.

---

## Checklist

- [ ] Please delete the labels section before submitting your issue
- [ ] I have described this issue in a way that is actionable (if possible)

### PR DESCRIPTION
**Pull request by seamusabshere**
https://example.com/shopify/auth/enable_cookies?shop=%27%22%3C/title%3E%3C/style%3E%3C/textarea%3E%3C/noscript%3E%3C/script%3E--%3E%3Cdtfy%3E%3Cscript%3Ealert(%27hello%20world%27)%3C/script%3E

Thanks to @nhusher

## Description

Fixes #1453

Screen Shot 2020-05-23 at 12 19 46 PM

### CODE REVIEW

File: packages/koa-shopify-auth/src/auth/create-enable-cookies.ts [hunk 1]
```diff
@@ -40,7 +40,7 @@ export default function createEnableCookies({
 
   <script>
     window.apiKey = "${apiKey}";
-    window.shopOrigin = "https://${shop}";
+    window.shopOrigin = "https://${encodeURIComponent(shop)}";
```
Comment (ricardotealdi) [inline packages/koa-shopify-auth/src/auth/create-enable-cookies.ts:43]: Should the `https://` be part of the `encodeURIComponent` call as well?

File: packages/koa-shopify-auth/src/auth/create-request-storage-access.ts [hunk 2]
```diff
@@ -39,7 +39,7 @@ export default function createRequestStorageAccess({
 
   <script>
     window.apiKey = "${apiKey}";
-    window.shopOrigin = "https://${shop}";
+    window.shopOrigin = "https://${encodeURIComponent(shop)}";
```
Comment (ricardotealdi) [inline packages/koa-shopify-auth/src/auth/create-request-storage-access.ts:42]: Same thing here.

File: packages/koa-shopify-auth/src/auth/create-enable-cookies.ts (hunk 1 above)
Comment (seamusabshere) [reply packages/koa-shopify-auth/src/auth/create-enable-cookies.ts:43]: i'm pretty sure "shop" always comes back as foo.myshopify.com

a couple years ago, @ragalie did this

https://github.com/Shopify/quilt/commit/3284be1e6cfe9165a2b23339d7f1bf8367ee9df7#diff-a15b7c0928331371cc40785dd637c644R16

which is a different way of fixing this, in a different place

Comment (theodoretan) [reply packages/koa-shopify-auth/src/auth/create-enable-cookies.ts:43]: 🤔 I think it's okay since we only want to encode the value coming from the user

Comment (ricardotealdi) [reply packages/koa-shopify-auth/src/auth/create-enable-cookies.ts:43]: That makes sense 👍

**Comment by seamusabshere** [comment]
In general, I have a feeling that this PR is wrong.

The `shop` parameter should be checked as soon as it's received from the user, not every single time it's used.

**Comment by ricardotealdi** [comment]
> The `shop` parameter should be checked as soon as it's received from the user, not every single time it's used.

Good point and I agree with you!

**Comment by jonpulsifer** [comment]
hi @seamusabshere have you heard about our white hat program over at https://hackerone.com/shopify? 💰

**Pull request merged**